sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+using Azure::Identity::AzureCliCredentialOptions;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+void ThrowIfNotSafeCmdLineInput(std::string const& input, std::string const& description)
+{
+  for (auto const c : input)
+  {
+    switch (c)
+    {
+      case ':':
+      case '/':
+      case '.':
+      case '-':
+      case '_':
+      case ' ':
+        break;
+
+      default:
+        if (!std::isalnum(c))
+        {
+          throw AuthenticationException(
+              "AzureCliCredential: Unsafe command line input found in " + description + ": "
+              + input);
+        }
+    }
+  }
+}
+} // namespace
+
+AzureCliCredential::AzureCliCredential(
+    std::string tenantId,
+    DateTime::duration cliProcessTimeout,
+    Core::Credentials::TokenCredentialOptions const& options)
+    : m_tenantId(std::move(tenantId)), m_cliProcessTimeout(std::move(cliProcessTimeout))
+{
+  static_cast<void>(options);
+  ThrowIfNotSafeCmdLineInput(m_tenantId, "TenantID");
+}
+
+AzureCliCredential::AzureCliCredential(AzureCliCredentialOptions const& options)
+    : AzureCliCredential(options.TenantId, options.CliProcessTimeout, options)
+{
+}
+
+AzureCliCredential::AzureCliCredential(TokenCredentialOptions const& options)
+    : AzureCliCredential(
+        AzureCliCredentialOptions{}.TenantId,
+        AzureCliCredentialOptions{}.CliProcessTimeout,
+        options)
+{
+}
+
+std::string AzureCliCredential::GetAzCommand(
+    std::string const& resource,
+    std::string const& tenantId) const
+{
+  ThrowIfNotSafeCmdLineInput(resource, "Resource");
+  std::string command = "az account get-access-token --output json --resource \"" + resource + "\"";
+
+  if (!tenantId.empty())
+  {
+    command += " --tenant \"" + tenantId + "\"";
+  }
+
+  return command;
+}
+
+namespace {
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context);
+}
+
+AccessToken AzureCliCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  try
+  {
+    auto const azCliResult = RunShellCommand(
+        GetAzCommand(
+            TokenCredentialImpl::FormatScopes(tokenRequestContext.Scopes, true, false), m_tenantId),
+        m_cliProcessTimeout,
+        context);
+
+    try
+    {
+      return TokenCredentialImpl::ParseToken(azCliResult, "accessToken", "expiresIn", "expiresOn");
+    }
+    catch (std::exception const&)
+    {
+      // Throw the az command output (error message)
+      // limited to 250 characters (250 has no special meaning).
+      throw std::runtime_error(azCliResult.substr(0, 250));
+    }
+  }
+  catch (std::exception const& e)
+  {
+    throw AuthenticationException(std::string("AzureCliCredential::GetToken(): ") + e.what());
+  }
+}
+
+namespace {
+class ShellProcess;
+class OutputPipe final {
+  friend class ShellProcess;
+
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_writeHandle = NULL;
```

Comment:
Use an RAII type for these handles - that way your code is exception safe in all cases.

Reply:
It is exception safe - these handles are in `ShellProcess` and `OutputPipe` classes which are RAII. Let me see what I can do - I tried it on early development stages but did not like the result - it was way too man lines, let me try once again and tell you.

Reply:
It's only exception safe because there are no opportunities in the current code to throw an exception in the constructor between when they are created and when the member variables are initialized. But that might change in the future and the future developer might not realize it.

That's why RAII wrappers on all handle and pointer types is (IMHO) a requirement. Implementing a complex class (like OutputPipe or ShellProcess with RAII techniques is a good start but it isn't foolproof.

Reply:
I am using `Azure::Core::_internal::UniqueHandle` now. I had to add another template parameter to it, which is a template template parameter with the default value of `Azure::Core::_internal::UniqueHandleHelper`. So, everything that was working before, still works.

I needed that second parameter, so the I can parameterize it with `<unnamed namespace>::UniqueHandleHelper`. Previously, there was no option for that - `UniqueHandleHelper` had to exist in `Azure::Core::_internal` namespace, and if I declared `UniqueHelper<HANDLE>` in the `Azure::Core::_internal` namespace, but in the Identity library, that symbol would be visible to linker, so if Core would ever define `UniqueHelper<HANDLE>` as well, there could be linker error because the same symbol would be present in both libraries.